libr/include/r_util/r_annotated_code.h

```diff
@@ -0,0 +1,123 @@
+
+#ifndef R2GHIDRA_ANNOTATEDCODE_H
+#define R2GHIDRA_ANNOTATEDCODE_H
+
+// #include <r_core.h>
+#include <r_types.h>
+#include <r_vector.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum r_syntax_highlight_type_t {
+	R_SYNTAX_HIGHLIGHT_TYPE_KEYWORD,
+	R_SYNTAX_HIGHLIGHT_TYPE_COMMENT,
+	R_SYNTAX_HIGHLIGHT_TYPE_DATATYPE,
+	R_SYNTAX_HIGHLIGHT_TYPE_FUNCTION_NAME,
+	R_SYNTAX_HIGHLIGHT_TYPE_FUNCTION_PARAMETER,
+	R_SYNTAX_HIGHLIGHT_TYPE_LOCAL_VARIABLE,
+	R_SYNTAX_HIGHLIGHT_TYPE_CONSTANT_VARIABLE,
+	R_SYNTAX_HIGHLIGHT_TYPE_GLOBAL_VARIABLE,
+} RSyntaxHighlightType;
+
+typedef enum r_code_annotation_type_t {
+	R_CODE_ANNOTATION_TYPE_OFFSET,
+	R_CODE_ANNOTATION_TYPE_SYNTAX_HIGHLIGHT,
+	// ...
+} RCodeAnnotationType;
+
+typedef struct r_code_annotation_t {
+	size_t start;
+	size_t end;
+	RCodeAnnotationType type;
+	union {
+		struct {
+			ut64 offset;
+		} offset;
+
+		struct {
+			RSyntaxHighlightType type;
+		} syntax_highlight;
+	};
+} RCodeAnnotation;
+
+typedef struct r_annotated_code_t {
+	char *code; // owned
+	RVector /*<RCodeAnnotation>*/ annotations;
+} RAnnotatedCode;
+
+/**
+* r_annotated_code_new() - Creates a new RAnnotatedCode structure and returns it's pointer.
+* @code: Literal code for which the RAnnotatedCode structure will be created .
+*
+* This functions creates a new RAnnotatedCode structure.
+* RAnnotatedCode.code will be initialized as the character array passed.
+* This will initialize RVector /*<RCodeAnnotation>*\/ annotations as well.
+* 
+* Return: Pointer to the new RAnnotatedCode structure created.
+*/
+R_API RAnnotatedCode *r_annotated_code_new(char *code);
+/**
+* r_annotated_code_free() - Deallocates *code.
+* @code: Pointer to a RAnnotatedCode.
+*
+* This functions deallocates memory allocated for *code.
+* 
+* Return: Nothing.
+*/
+R_API void r_annotated_code_free(RAnnotatedCode *code);
+/**
+* r_annotated_code_add_annotation() - Inserts *annotation in *code.
+* @code: Pointer to a RAnnotatedCode.
+* @annotation: Pointer to a annotation.
+*
+* This functions inserts the annotation represented by the pointer 'annotation' to the vector
+* of annotations in the RAnnotatedCode represented by 'code'. To be more precise,
+* annotation will be added to code->annotations, which is a RVector /*<RCodeAnnotation>*\/ annotations.
+* 
+* Return: Nothing.
+*/
+R_API void r_annotated_code_add_annotation(RAnnotatedCode *code, RCodeAnnotation *annotation);
+/**
+* r_annotated_code_annotations_in() - Returns all annotations with range that contains the given offset.
+* @code: Pointer to a RAnnotatedCode.
+* @offset: Offset.
+*
+* Creates an RPVector and inserts the pointers to all annotations in which 
+* annotation->start <= offset < annotation->end.
+* This RPVector can have annotations of the type R_CODE_ANNOTATION_TYPE_OFFSET and 
```

Comment:
I'm not sure this is really necessary to be said... these are anyway the only type of annotation we have right now.

Reply:
I have explained the types of annotation in another comment. Or were you talking about something else here?

Reply:
I'm saying that if there are just 2 types of annotations and this function returns all kinds of annotations, then there is no need to list each possible type of annotations that can be returned in the comment.

Reply:
@ret2libc I wrote all these replies without seeing any of your reply to the above comments. It didn't refresh automatically. Sorry for that.

Yes, I understand your point. I will remove this part from the comment.